File: Algorithm Program/22-23/fall/lecture/math/7-combinatorics-mod/factorial.py
# https://ideone.com/rskzqd
from random import randint
from sys import setrecursionlimit

setrecursionlimit(10**6)

MOD = 10**9 + 7


def factorial(n: int) -> int:
    res = 1
    for i in range(1, n + 1):
        res *= i
        res %= MOD
    return res


def factorial_recursive(n: int) -> int:
    if not n:
        return 1
    return n * factorial_recursive(n - 1) % MOD


def print_factorials(bound: int) -> int:
    n_examples = 10
    for i in range(n_examples):
        n = randint(0, bound)
        print(f"{n}!:")
        print(f"{factorial(n)} (iterative)")
        print(f"{factorial_recursive(n)} (recursive)")


bound = 10**5
print_factorials(bound)

File: Algorithm Program/22-23/fall/lecture/math/7-combinatorics-mod/factorial.cpp
// https://ideone.com/OLseOi
#include <bits/stdc++.h>
using namespace std;

const int mod = 1e9 + 7;

int mul(int a, int b) {
    return 1ll * a * b % mod;
}

int factorial(int n) {
    int res = 1;
    for (int i = 1; i <= n; i++) {
        res = mul(res, i);
    }
    return res;
}

int factorial_recursive(int n) {
    if (!n)
        return 1;
    return mul(n, factorial_recursive(n - 1));
}

void print_factorials(int bound) {
    int n_examples = 10;
    for (int i = 0; i <= n_examples; i++) {
        int n = rand() % (bound + 1);
        cout << n << "!:\n";
        cout << "\t" << factorial(n) << " (iterative)\n";
        cout << "\t" << factorial_recursive(n) << " (recursive)\n";
    }
}

int main() {
    int bound = 1e7;
    print_factorials(bound);
}